server/src/main/java/org/elasticsearch/search/fetch/subphase/FetchFieldsPhase.java

```diff
@@ -56,23 +58,47 @@ public FetchSubPhaseProcessor getProcessor(FetchContext fetchContext) {
-        final FieldFetcher fieldFetcher = fetchFieldsContext == null ? null
-            : fetchFieldsContext.fields() == null ? null
-            : fetchFieldsContext.fields().isEmpty() ? null
-            : FieldFetcher.create(searchExecutionContext, fetchFieldsContext.fields());
+        final FieldFetcher fieldFetcher;
+        if (fetchFieldsContext != null && fetchFieldsContext.fields() != null && fetchFieldsContext.fields().isEmpty() == false) {
+            final List<FieldAndFormat> fetchFields = fetchFieldsContext.fields()
+                .stream()
+                .filter(
+                    fieldAndFormat -> searchExecutionContext.isMetadataField(fieldAndFormat.field) == false
+                        || searchExecutionContext.getFieldType(fieldAndFormat.field).isStored() == false
+                )
+                .toList();
+            fieldFetcher = FieldFetcher.create(searchExecutionContext, fetchFields);
+        } else {
+            fieldFetcher = FieldFetcher.create(searchExecutionContext, Collections.emptyList());
+        }
 
+        final Set<FieldAndFormat> additionalMetadataFields = new HashSet<>();
+        if (fetchFieldsContext != null && fetchFieldsContext.fields() != null && fetchFieldsContext.fields().isEmpty() == false) {
+            final List<FieldAndFormat> storedMetadataFieldsFromFetchContext = fetchFieldsContext.fields()
+                .stream()
+                .filter(
+                    fieldAndFormat -> searchExecutionContext.isMetadataField(fieldAndFormat.field)
+                        && searchExecutionContext.getFieldType(fieldAndFormat.field).isStored()
+                )
+                .toList();
+            additionalMetadataFields.addAll(storedMetadataFieldsFromFetchContext);
+        }
+        final Set<FieldAndFormat> metadataFields = new HashSet<>(DEFAULT_METADATA_FIELDS);
+        metadataFields.addAll(additionalMetadataFields);
         final FieldFetcher metadataFieldFetcher;
         if (storedFieldsContext != null
             && storedFieldsContext.fieldNames() != null
             && storedFieldsContext.fieldNames().isEmpty() == false) {
-            final Set<FieldAndFormat> metadataFields = new HashSet<>(DEFAULT_METADATA_FIELDS);
             for (final String storedField : storedFieldsContext.fieldNames()) {
                 final Set<String> matchingFieldNames = searchExecutionContext.getMatchingFieldNames(storedField);
                 for (final String matchingFieldName : matchingFieldNames) {
                     if (SourceFieldMapper.NAME.equals(matchingFieldName) || IdFieldMapper.NAME.equals(matchingFieldName)) {
                         continue;
                     }
                     final MappedFieldType fieldType = searchExecutionContext.getFieldType(matchingFieldName);
+                    if (Regex.isSimpleMatchPattern(storedField) && IgnoredSourceFieldMapper.NAME.equals(matchingFieldName)) {
```

Comment:
This is required to avoid fetching `_ignored_source` when requested via a wildcard `*`